### PR TITLE
LPD-35757 Prevent InstanceScoped JS Client Extension from being applied at any other scope

### PR DESCRIPTION
**Pull request by thiago-buarqque**
# Context
We now have a way to inject a globalJS CET into all pages of a instance. See https://liferay.atlassian.net/browse/LPD-33747 for more info

Original PR: https://github.com/liferay-platform-experience/liferay-portal/pull/706

## Original description for this PR

Task: https://liferay.atlassian.net/browse/LPD-35757
Story: https://liferay.atlassian.net/browse/LPD-35637

## Motivation
There is no benefit of allowing a JS client extension, that is applied to the entire instance, to be applied to the page level. 

## Goal
Hide instance scoped cxs when a user is selecting a globalJS cx to a page.

### Steps to reproduce

1. Deploy a globalJS client extension with instance scope. You can deploy the one in the sample workspace
2. Go to `Site builder > pages > configuration > JavaScript > in page head` and assert that the client extension is not listed there.
3. Enable the feature flag LPD-30371
4. Repeat step 2

### CODE REVIEW

**Comment by thiago-buarqque** [comment]
ci:test:relevant

File: modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java [hunk 1]
```diff
@@ -31,8 +31,16 @@ public CET addCET(
 	public CET getCET(long companyId, String externalReferenceCode);
 
 	public List<CET> getCETs(
-			long companyId, String keywords, String type, Pagination pagination,
-			Sort sort)
+			long companyId, boolean excludeInstanceScopedCETs, String keywords,
+			String type, Pagination pagination, Sort sort)
+		throws PortalException;
```
Comment (dsanz) [inline modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java:36]: If we change an existing method signature, all providers and consumers of this interface are forced to change. 

An alternative is to keep the existing method untouched and add a new one:
``` 
public List<CET> getCETs(
			long companyId, boolean excludeInstanceScopedCETs, String keywords,
			String type, Pagination pagination, Sort sort)
```
This way, we can turn the implementation of the old one as a special case of invocation of the new one. We can even do it in the interface itself as a default method, so that it'll be available for the implementations:

``` 
public defaul List<CET> getCETs(
			long companyId, String keywords, String type, Pagination pagination,
			Sort sort) {
          return getCETs(companyId, false, keywords, type, pagination, sort);
}
```
Now, `CETManagerImpl` would just implement the new method as you do [here](https://github.com/liferay-frontend/liferay-portal/pull/4416/files#diff-438b6f11f78bd05fac56fb7890e5e939bfe147f1c09d74b4784cf3ee6fdcdb56R104), and there is no need for it to implement the old one.

It's also possible to implement both methods in the implementation class. The point is that most usages in the consumers don't need this flag so both methods shall be made available. This way, we are not breaking potential client extension types implemented out there.

File: modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java [hunk 2]
```diff
@@ -31,8 +31,16 @@ public CET addCET(
...
+
+	public List<CET> getCETs(long companyId, String type)
+		throws PortalException;
```
Comment (dsanz) [inline modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java:39]: This is a new method, which is nice for some usages. However, its equivalent getCETsCount does not exist as it still takes a list of keywords. This makes the API a bit inconsistent because interface breaks the parity between the `getCETs` and `getCETsCount` counterparts, where we had:

- `getCETs(a, b, c, pagination, sort)`
- `getCETsCount(a, b, c)`   (parameters before pagination/sort remain the same)

Which becomes:

- `getCETs(a, b, c, d, pagination, sort)`
- `getCETsCount(a, b, c, d)`  (so far so good)
- `getCETs(a, b)`
- `getCETsCount(a, b, c)`  (this one breaks parity)

This is a minor nit, nevertheless, is there any reason to keep interface this way I am missing?

**Review by dsanz — CHANGES_REQUESTED** [review]
Hey Thiago, pls take a look to the proposal I'm making and let me know what fo you think. Thank you

File: modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java (hunk 2 above)
Comment (thiago-buarqque) [reply modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java:39]: Got it, you have a point. I'm thinking about keeping this method, so I will add a counter version for it

Comment (thiago-buarqque) [reply modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java:39]: @dsanz  Actually, there is only [one usage](https://github.com/liferay/liferay-portal/blob/6ce0b9eb049e99a50285a324a3bd4afd76d340ab/modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/frontend/data/set/provider/CETFDSDataProvider.java#L66) for getCETsCount. The shorter method is handy so the client code doesn't have to pass in a bunch of null and default values

Comment (thiago-buarqque) [reply modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java:39]: Do you think keeping this method without its count version is fine for now?

Comment (thiago-buarqque) [reply modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/manager/CETManager.java]: I update the PR, kept the way it is because, until now, there is no need to have a count version of it.

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/manager/CETManagerImpl.java [hunk 3]
```diff
@@ -209,11 +218,25 @@ private Map<String, CET> _getCETsMap(long companyId) {
...
-	private boolean _isInclude(CET cet, String keywords, String type) {
+	private boolean _isInclude(
+		CET cet, boolean excludeInstanceScopedCETs, String keywords,
+		String type) {
+
 		if (Validator.isNotNull(type) && !Objects.equals(type, cet.getType())) {
 			return false;
 		}
 
+		if (excludeInstanceScopedCETs &&
```
Comment (izaera) [inline modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/manager/CETManagerImpl.java:229]: :thinking: I don't think it makes sense to change the `CETManager` API by adding a new concept like scope if it is not going to be applied to all CX types.

Either we:
-  Promote the scope to be generalized to all CXs (and in that case it would have to be stored in the DB, not in the `typeSettings`) or
- We don't do anything to `CETManager` and implement the `excludeInstanceScopedCETs`  filtering at the higher level (that would be `CETItemSelectorViewDescriptor`).

**Review by izaera — CHANGES_REQUESTED** [review]
I've left one comment regarding the change in the `CETManager` API. Other than that the rest seems OK to me :+1: .

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/manager/CETManagerImpl.java (hunk 3 above)
Comment (dsanz) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/manager/CETManagerImpl.java:229]: mmm I'm having a second thought, I believe you're right Iván. I misunderstood the role of this flag.

The `instance scope CX` is a concept that only applies to one type of CET. That's changes things, as it's not a general purpose flag. 

IINM, flag influences in 2 aspects:

1.  the way that GlobalJS entries are injected in the response, that is, rather than choosing the page where it'll be added, it's just added unconditionally. 
2.  filter a list of GlobalJS entries for user selection. Usually, filter by type is enough, but not in this case as they need a subset of the GlobalJS, those not marked as "instance scoped".

Essentially, this flag is no different from things we already have in other CX types. For example Editor Config Contributor CX type declares 3 properties that dictate when such entries are being executed to contribute configurations to a given editor instance. Difference is that we don't offer users a filtered list of Editor Config Contributor entries for selection.

I do agree this need does not have to be promoted into the framework itself: any additional filtering that depends on per-type specificities should not manifest in the main Manager API. 

I would defend this being added to the framework as long as one of the following conditions are met:

- The new characteristic is common to all CET types, with no exception.
- Framework provides an agnostic way of adding extra filtering to the CX, by accepting a consumer-provided predicate function that is applied to all items in a way that can't be done from outside. This way different consumers would implement specific filtering functions. 
 
Nevertheless, as this filtering is an "additional last step" after the list has been returned, I don't see any benefit of adding this to the framework either. A different thing would be if such predicate had to be applied in the middle of the query, but this does not seem the case: first filter by type, then exclude those with "instance scope".

To sum up, CETManager should not deal with this, as otherwise we would keep adding specific (per type) methods to a general-purpose interface.

Thiago, could you please try to filter the list of GlobalJS CX entries in the `CETItemSelectorViewDescriptor.getSearchContainer() `as Iván suggests? This looks a much better place for this, there's no need to change any API.

Thank you!

Comment (izaera) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/manager/CETManagerImpl.java:229]: BTW, you should also add a getter in the CX type to be able to do the filtering and so that everybody knows that field exits. That would be this interface -> https://github.com/liferay/liferay-portal/blob/90fdc04e1bac43119571af253a2b09b685a33234/modules/apps/client-extension/client-extension-type-api/src/main/java/com/liferay/client/extension/type/GlobalJSCET.java#L18, its implementation, etc.

**Comment by ethib137** [comment]
Hey @dsanz and @izaera , thanks for the reviews. You should also know that we do want other types of CX's to be applied at the instance scope: themeCSS, CSS, spritemap, and themeFavicon. All of these would provide value by being applied at the instance level or to the admin interfaces. Just one more thing to be aware of.

File: modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/manager/CETManagerImpl.java (hunk 3 above)
Comment (thiago-buarqque) [reply modules/apps/client-extension/client-extension-type-impl/src/main/java/com/liferay/client/extension/type/internal/manager/CETManagerImpl.java:229]: Hey guys. I believe you are right. I did this because we are going to add this concept to other types of client extensions. But i don't think it will be all of them. I will update the PR with the suggested approach

**Comment by thiago-buarqque** [comment]
Hey @ethib137 @dsanz @izaera I have updated the PR with the simpler approach. Please let me know if I missed something

**Comment by izaera** [comment]
Looks good to me now :+1: Thx

**Review by izaera — APPROVED** [review]

**Review by dsanz — APPROVED** [review]
LGTM ;) thanks Thiago!

**Comment by dsanz** [comment]
ci:forward

**Comment by thiago-buarqque** [comment]
ci:test:stable

**Comment by thiago-buarqque** [comment]
ci:forward:force
